### PR TITLE
Support decimal create with leading zero

### PR DESCRIPTION
**Pull request by kisbg**
JerryScript-DCO-1.0-Signed-off-by: bence gabor kis kisbg@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/parser/js/js-lexer.c [hunk 1]
```diff
@@ -1429,11 +1429,36 @@ lexer_parse_number (parser_context_t *context_p) /**< context */
...
     }
     else if (source_p[1] >= LIT_CHAR_8
              && source_p[1] <= LIT_CHAR_9)
     {
+      if (context_p->status_flags & PARSER_IS_STRICT)
+      {
+        parser_raise_error (context_p, PARSER_ERR_OCTAL_NUMBER_NOT_ALLOWED);
+      }
+#if ENABLED (JERRY_ESNEXT)
+      lexer_check_numbers (context_p, &source_p, source_end_p, LIT_CHAR_9, true);
```
Comment (zherczeg) [inline jerry-core/parser/js/js-lexer.c:1452]: Why `is_legacy` is true?

File: jerry-core/parser/js/js-lexer.c [hunk 2]
```diff
@@ -1429,11 +1429,36 @@ lexer_parse_number (parser_context_t *context_p) /**< context */
...
+      if (source_p[0] >= LIT_CHAR_8
+          && source_p[0] <= LIT_CHAR_9)
+      {
+#if ENABLED (JERRY_ESNEXT)
+        lexer_check_numbers (context_p, &source_p, source_end_p, LIT_CHAR_9, false);
+        context_p->token.extra_value = LEXER_NUMBER_DECIMAL;
+#else
+        parser_raise_error (context_p, PARSER_ERR_INVALID_NUMBER);
+#endif /* ENABLED (JERRY_ESNEXT) */
+      }
```
Comment (zherczeg) [inline jerry-core/parser/js/js-lexer.c:1442]: Is it possible to combine the two cases somehow to reduce code size?

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: jerry-core/parser/js/js-lexer.c (hunk 1 above)
Comment (kisbg) [reply jerry-core/parser/js/js-lexer.c]: The following test-cases need to throw a syntax error: 08_0.

File: jerry-core/parser/js/js-lexer.c (hunk 2 above)
Comment (zherczeg) [reply jerry-core/parser/js/js-lexer.c:1442]: There was no reply to this question. The `else` in line 1444 seems es5.1 specific, and all esnext cases could be merged here.

Comment (kisbg) [reply jerry-core/parser/js/js-lexer.c:1442]: I am sorry for not answering to your question. I Simplified the code.

File: jerry-core/parser/js/js-lexer.c [hunk 3]
```diff
@@ -1429,11 +1429,17 @@ lexer_parse_number (parser_context_t *context_p) /**< context */
       }
 
       lexer_check_numbers (context_p, &source_p, source_end_p, LIT_CHAR_7, true);
-    }
-    else if (source_p[1] >= LIT_CHAR_8
-             && source_p[1] <= LIT_CHAR_9)
-    {
-      parser_raise_error (context_p, PARSER_ERR_INVALID_NUMBER);
+
+      if (source_p[0] >= LIT_CHAR_8
```
Comment (zherczeg) [inline jerry-core/parser/js/js-lexer.c:1433]: Only one more thing: you must check the buffer end to avoid buffer overread

**Review by zherczeg — CHANGES_REQUESTED** [review]
I like the merge, considerably simpler.

**Comment by kisbg** [comment]
@zherczeg  Thank you for the review!

File: jerry-core/parser/js/js-lexer.c [hunk 4]
```diff
@@ -1429,11 +1429,17 @@ lexer_parse_number (parser_context_t *context_p) /**< context */
       }
 
       lexer_check_numbers (context_p, &source_p, source_end_p, LIT_CHAR_7, true);
-    }
-    else if (source_p[1] >= LIT_CHAR_8
-             && source_p[1] <= LIT_CHAR_9)
-    {
-      parser_raise_error (context_p, PARSER_ERR_INVALID_NUMBER);
+
+      if (source_p <= source_end_p && source_p[0] >= LIT_CHAR_8
```
Comment (zherczeg) [inline jerry-core/parser/js/js-lexer.c:1433]: This syntax is invalid. Please put every check into different line. Also `source_p < source_end_p`

File: jerry-core/parser/js/js-lexer.c [hunk 5]
```diff
@@ -1429,11 +1429,17 @@ lexer_parse_number (parser_context_t *context_p) /**< context */
...
+          && source_p[0] <= LIT_CHAR_9)
+      {
+#if ENABLED (JERRY_ESNEXT)
+        lexer_check_numbers (context_p, &source_p, source_end_p, LIT_CHAR_9, true);
+        context_p->token.extra_value = LEXER_NUMBER_DECIMAL;
+#else
```
Comment (rerobika) [inline jerry-core/parser/js/js-lexer.c:1439]: Missing comment `!ENABLED (JERRY_ESNEXT)`

**Review by zherczeg — APPROVED** [review]
LGTM

File: tests/jerry/es.next/decimal-with-leading-zero.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    assert (true)
+  }
+}
+
+// invalid strict test-cases
+invalid_strict_cases ("09")
+invalid_strict_cases ("01239")
+
+// invalid to create bigint with decimal with leading zero
+function invalid_cases (string)
```
Comment (zherczeg) [inline tests/jerry/es.next/decimal-with-leading-zero.js:36]: Is this function used?

Comment (kisbg) [reply tests/jerry/es.next/decimal-with-leading-zero.js:36]: Whoops! I called the wrong function. Thank you for find this error!

**Review by zherczeg — APPROVED** [review]
LGTM

**Review by rerobika — APPROVED** [review]
LGTM

**Pull request merged**